src/core/FdirSourceFileProvider.ts

```diff
@@ -0,0 +1,137 @@
+    matchPath: MatchPathAsync;
+
+    constructor(configFileName: NormalizedPath, private rootDirs: string[]) {
+        this.parsedCommandLine = ts.getParsedCommandLineOfConfigFile(
+            configFileName,
+            {},
+            {
+                getCurrentDirectory: process.cwd,
+                fileExists: fs.existsSync,
+                useCaseSensitiveFileNames: true,
+                readFile: path => fs.readFileSync(path, 'utf-8'),
+                readDirectory: () => {
+                    // this is supposed to be the recursive file walk.
+                    // since we don't care about _actually_ discovering files,
+                    // only about parsing the config's compilerOptions
+                    // (and tracking the "extends": fields across multiple files)
+                    // we short circuit this.
+                    return [];
+                },
+                onUnRecoverableConfigFileDiagnostic: diagnostic => {
+                    console.error(diagnostic);
+                    process.exit(1);
+                },
+            }
+        );
+
+        this.matchPath = createMatchPathAsync(
+            this.parsedCommandLine.options.baseUrl,
+            this.parsedCommandLine.options.paths
+        );
+    }
+
+    async getSourceFiles(searchRoots?: string[]): Promise<string[]> {
+        const allRootsDiscoveredFiles: string[][] = await Promise.all(
+            (searchRoots || this.rootDirs).map(
+                (rootDir: string) =>
+                    new fdir()
+                        .glob(
+                            this.parsedCommandLine.options.allowJs
+                                ? '**/!(.d)*@(.js|.ts)'
+                                : '**/*!(.d)*.ts'
+                        )
+                        .withFullPaths()
+                        .crawl(rootDir)
+                        .withPromise() as Promise<string[]>
+            )
+        );
+
+        return [...new Set<string>(allRootsDiscoveredFiles.reduce((a, b) => a.concat(b), []))];
+    }
+
+    async getImportsForFile(filePath: string): Promise<string[]> {
+        const fileInfo = ts.preProcessFile(await readFile(filePath, 'utf-8'), true, true);
+        return fileInfo.importedFiles.map(importedFile => importedFile.fileName);
+    }
+
+    async resolveImportFromFile(
+        importer: string,
+        importSpecifier: string
+    ): Promise<string | undefined> {
+        if (importSpecifier.startsWith('.')) {
+            // resolve relative and check extensions
+            return await checkExtensions(
+                path.join(path.dirname(importer), importSpecifier),
+                ALLOWED_EXTENSIONS
+            );
+        } else {
+            // resolve with tsconfig-paths (use the paths map, then fall back to node-modules)
+            return await new Promise((res, rej) =>
+                this.matchPath(
+                    importSpecifier,
+                    undefined,

```

Comment:
I think we might be able to cache the result of getAllFiles() and resolve from that list. matchPath provides a way to override its fs access, so we could avoid redundent disc access here to squeeze out a few more milliseconds. I think my last perf checkpoints had this walk as a pretty small amount though.

Reply:
Resolving this as work to do in a follow-up